tests/components/color_extractor/test_service.py

```diff
@@ -0,0 +1,315 @@
+        ATTR_ENTITY_ID: LIGHT_ENTITY,
+    }
+
+    # Don't let the URL not being allowed sway our exception test
+    hass.config.allowlist_external_urls.add("http://example.com/images/")
+
+    # Mock the HTTP Response with an HTTPError
+    aioclient_mock.get(url=service_data[ATTR_URL], exc=aiohttp.ClientError)
+
+    await _async_load_color_extractor_url(hass, service_data)
+
+    # Light has not been modified due to failure
+    state = hass.states.get(LIGHT_ENTITY)
+    assert state
+    assert state.state == STATE_OFF
+
+
+async def test_url_error(hass, aioclient_mock):
+    """Test that a HTTP Error (non 200) doesn't turn light on."""
+    service_data = {
+        ATTR_URL: "http://example.com/images/logo.png",
+        ATTR_ENTITY_ID: LIGHT_ENTITY,
+    }
+
+    # Don't let the URL not being allowed sway our exception test
+    hass.config.allowlist_external_urls.add("http://example.com/images/")
+
+    # Mock the HTTP Response with a 400 Bad Request error
+    aioclient_mock.get(url=service_data[ATTR_URL], status=400)
+
+    await _async_load_color_extractor_url(hass, service_data)
+
+    # Light has not been modified due to failure
+    state = hass.states.get(LIGHT_ENTITY)
+    assert state
+    assert state.state == STATE_OFF
+
+
+@patch(
+    "builtins.open",
+    mock_open(read_data=base64.b64decode(load_fixture("color_extractor_file.txt"))),
+    create=True,
+)
+def _get_file_mock(file_path):
+    """Convert file to BytesIO for testing due to PIL UnidentifiedImageError."""
+    _file = None
+
+    with open(file_path) as file_handler:
+        _file = io.BytesIO(file_handler.read())
+
+    _file.name = "color_extractor.jpg"
+    _file.seek(0)
+
+    return _file
+
+
+@patch("os.path.isfile", Mock(return_value=True))
+@patch("os.access", Mock(return_value=True))
+async def test_file(hass):
+    """Test that the file only service reads a file and translates to light RGB."""
+    service_data = {
+        ATTR_PATH: "/opt/image.png",
+        ATTR_ENTITY_ID: LIGHT_ENTITY,
+        # Standard light service data which we pass
+        ATTR_BRIGHTNESS_PCT: 100,
+    }
+
+    # Add our /opt/ path to the allowed list of paths
+    hass.config.allowlist_external_dirs.add("/opt/")
+
+    await async_setup_component(hass, DOMAIN, {})
+    await hass.async_block_till_done()
+
+    # Verify pre service check
+    state = hass.states.get(LIGHT_ENTITY)
+    assert state
+    assert state.state == STATE_OFF
+
+    # Mock the file handler read with our 1x1 base64 encoded fixture image
+    with patch("homeassistant.components.color_extractor._get_file", _get_file_mock):
+        await hass.services.async_call(DOMAIN, SERVICE_TURN_ON, service_data)
+        await hass.async_block_till_done()
+
+    state = hass.states.get(LIGHT_ENTITY)
+
+    assert state
+
+    # Ensure we turned it on
+    assert state.state == STATE_ON
+
+    # And set the brightness
+    assert state.attributes[ATTR_BRIGHTNESS] == 255
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6372789)